### PR TITLE
Changes to avoid N+1 and add integration specs

### PR DESCRIPTION
**Pull request by redmarverick**
### Add Eager Loading to comments, likes and posts to better the performance

<details>
<summary>Posts queries</summary>

### Before the changes

b4 n+1

### After the changes to avoid N+1

The log below shows that the associated data for the post, such as comments and likes, are being loaded efficiently without causing additional queries (thanks to the caching mechanism as well).
This optimization helps in avoiding the N+1 query problem, which can lead to better performance by reducing the number of database queries executed. The controller action is now making use of eager loading and caching.

after n+1

</details>

<details>
<summary>Comments queries</summary>

### Comments queries before

b4 n+1 comments

### Comments queries after changes to avoid N+1

Just as expected, the eager loading for comments and likes is working on UserPostsController. The log shows that the associated comments and likes for the posts on the page are being loaded efficiently without causing additional queries.

after n+1 comments

</details>

### Created Spec tests:

- User index page:

1. I can see the username of all other users.
2. I can see the profile picture for each user.
3. I can see the number of posts each user has written.
4. When I click on a user, I am redirected to that user's show page.
5. 

- user show page:

1. I can see the user's profile picture.
2. I can see the user's username.
3. I can see the number of posts the user has written.
4. I can see the user's bio.
5. I can see the user's first 3 posts.
6. I can see a button that lets me view all of a user's posts.
7. When I click a user's post, it redirects me to that post's show page.
8. When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:

1. I can see the user's profile picture.
2. I can see the user's username.
3. I can see the number of posts the user has written.
4. I can see a post's title.
5. I can see some of the post's body.
6. I can see the first comments on a post.
7. I can see how many comments a post has.
8. I can see how many likes a post has.
9. I can see a section for pagination if there are more posts than fit on the view.
10. When I click on a post, it redirects me to that post's show page.

- Post show page:

1. I can see the post's title.
2. I can see who wrote the post.
3. I can see how many comments it has.
4. I can see how many likes it has.
5. I can see the post body.
6. I can see the username of each commentor.
7. I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/user_posts_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  it 'I can see post information and click on a post' do
+    visit user_posts_path(user_id)
+
+    expect(page).to have_content(@user.name)
+    expect(page).to have_content("Posts: #{@user.posts_counter}")
+
+    first(:css, 'a', text: /Post Title/).click
+    expect(current_path).to include('/users/250/posts')
+    expect(page).to have_content('Post Title', minimum: 1)
+    expect(page).to have_content('Post Title', minimum: 1)
```
Comment (youmari) [inline spec/features/user_posts_index_spec.rb:32]: - [x] As the requirements say you should add the following tests in order to complete this project 
- Tests that assert you can see some of the post's body.
- Tests that assert you can see a section for pagination if there are more posts than fit on the view.
- Tests that assert you can see the user's profile picture.
- Tests that redirect to the post's show page when you click on a post. it should be something like this '/users/250/posts/1'

File: spec/features/user_posts_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    expect(page).to have_content(@post.title)
+    expect(page).to have_content("Comments: 3")
+    expect(page).to have_content("Likes: 0")
+    expect(page).to have_content(@post.text)
+
+    @comments.each do |comment|
+      expect(page).to have_content(comment.author.name)
+      expect(page).to have_content(comment.text)
+    end
+  end
```
Comment (youmari) [inline spec/features/user_posts_show_spec.rb]: - [ ] As the requirements say you should add the following tests in order to complete this project

- Tests that assert you can see who wrote the post.
- Tests that assert you can see the post body.

**Review by youmari — CHANGES_REQUESTED** [review]
Hi @redmarverick , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
✔️ Great commit messages
✔️ PRO PR
✔️ No linter error 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_posts_index_spec.rb (hunk 1 above)
Comment (redmarverick) [reply spec/features/user_posts_index_spec.rb:32]: correction done

**Review by cynthiainga — APPROVED** [review]
## STATUS APPROVED ✔️ 🎉 ✔️ 

Hi @redmarverick, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- [ ] No linters errors ✔️ 
- [ ] Correct Gitflow ✔️ 
- [ ] Descriptive PR ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**